feat(e2e-webdriverio): add e2e plugin for WebdriverIO

christian-bromann opened the pull request:
<!-- Please don't delete this template -->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Underlying tools
- [ ] Other, please describe:

This PR adds a new e2e option to run tests using [WebdriverIO](https://webdriver.io/):

![demo](https://user-images.githubusercontent.com/731337/81921802-774af900-95db-11ea-98dd-29907c11be22.gif)

I am happy to be help triaging and fixing all issues that users might create using this plugin. I am the WebdriverIO maintainer and invested in the usability of this.

<!--
Note:
When submitting documentation PRs, please target the `master` branch (https://cli.vuejs.org) or `next` branch (https://next.cli.vuejs.org)
When submitting coding PRs, please target the `dev` branch.
-->

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**Other information:**

sodatea commented:
Thanks for the pull request!

As we are very close to the next scheduled feature release, I'm not able to review such a big piece of code in the next few days.

I'll take a look later and hope to fit it into the next feature release (scheduled in June).

christian-bromann commented:
Any updates on this?

sodatea commented on packages/@vue/cli-plugin-e2e-webdriverio/README.md:
````diff
@@ -0,0 +1,79 @@
+## Installing in an Already Created Project
+
+``` sh
+vue add e2e-nightwatch
````
```suggestion
vue add e2e-webdriverio
```

sodatea replied:
You also need to edit https://github.com/vuejs/vue-cli/blob/dev/packages/@vue/cli-shared-utils/lib/pluginResolution.js#L5 to add this plugin to the official plugin list.

Better add a note here that for users of older CLI versions, they may need to run `vue add @vue/e2e-webdriverio`, because the official plugin list was hard-coded.

sodatea commented on packages/@vue/cli-plugin-e2e-webdriverio/package.json:
```diff
@@ -0,0 +1,41 @@
+    "@wdio/sauce-service": "^6.1.9",
+    "@wdio/spec-reporter": "^6.1.9",
+    "@wdio/sync": "^6.1.8",
+    "chromedriver": "^81.0.0",
```
Considering chromedriver updates its major version number quite frequently, could you please make it a peer dependency, like what we did in the nightwatch plugin?

https://github.com/vuejs/vue-cli/blob/dev/packages/%40vue/cli-plugin-e2e-nightwatch/package.json
https://github.com/vuejs/vue-cli/blob/dev/packages/%40vue/cli-plugin-e2e-nightwatch/prompts.js
https://github.com/vuejs/vue-cli/blob/dev/packages/@vue/cli-plugin-e2e-nightwatch/generator/index.js

christian-bromann commented:
@sodatea thanks for the review. I updated the PR according to your suggestions.

christian-bromann commented:
@sodatea please let me know if there is anything missing.

sodatea commented on packages/@vue/cli-plugin-e2e-webdriverio/package.json:
```diff
@@ -0,0 +1,43 @@
+  "peerDependencies": {
+    "chromedriver": "*"
+  },
+  "devDependencies": {}
```
Need to add `chromedriver` and `geckodriver` to `devDependencies` too.

sodatea commented on packages/@vue/cli-plugin-e2e-webdriverio/generator/index.js:
```diff
@@ -0,0 +1,33 @@
+const { installedBrowsers } = require('@vue/cli-shared-utils')
+
+module.exports = (api, { webdrivers }) => {
```
Need to copy-paste the `prompts.js` from the nightwatch plugin, too.

christian-bromann replied:
It seems that the prompt actually comes from `packages/@vue/cli/lib/promptModules/e2e.js`. Changing the `prompt.js` in the plugin package has no effect.

sodatea replied:
`promptModules/*` are used on `vue create`, `prompt.js` is for `vue add` and `vue invoke`.

sodatea commented on packages/@vue/cli/lib/promptModules/e2e.js:
```diff
@@ -25,6 +25,11 @@ module.exports = cli => {
+      {
+        name: 'WebdriverIO (WebDriver/DevTools based)',
+        value: 'webdriverio',
+        short: 'WebdriverIO'
```
Should chnage line 39 from `when: answers => answers.e2e === 'nightwatch'` to `when: answers => answers.e2e === 'nightwatch' || answers.e2e === 'webdriverio'` (I can't directly add review comment on that line in the GitHub web ui)

sodatea commented on packages/@vue/cli-plugin-e2e-webdriverio/index.js:
```diff
@@ -0,0 +1,71 @@
+      }
+
+      const configFile = !args.remote
+        ? path.join(process.cwd(), 'tests', 'wdio.local.conf.js')
+        : path.join(process.cwd(), 'tests', 'wdio.sauce.conf.js')
```
The `tests` part seems extraneous? As the config files are at the project root according to the templates.

sodatea commented on packages/@vue/cli-plugin-e2e-webdriverio/generator/template/tests/e2e/specs/app.spec.js:
```diff
@@ -0,0 +1,8 @@
+describe('Vue.js app', () => {
+  it('should open and render', () => {
+    App.open()
+    expect(App.heading).toHaveText('Welcome to Your Vue.js App')
```
```suggestion
    expect(App.heading).toHaveText('Welcome to Your Vue.js <%- hasTS ? '+ TypeScript ' : '' %>App')
```

sodatea commented on packages/@vue/cli-plugin-e2e-webdriverio/generator/template/wdio.local.conf.js:
```diff
@@ -0,0 +1,21 @@
+        : []
+    }
+  }]
+}
```
Could you add a firefox example if we are going to allow choosing the browser on project creation?

christian-bromann replied:
Added Firefox browser to the setup

christian-bromann commented:
@sodatea thanks for the review. I addressed the comments and also added WebdriverIO to the docs (english only). The build is failing due to connectivity issues, let's rerun them.

sodatea commented:
Seems the build isn't failing because of network issues. But somehow yarn cannot resolve the `*` version range specified in a workspace package's dependencies. Let's copy the ones from the nightwatch plugin as a workaround https://github.com/vuejs/vue-cli/blob/d319007ffa702986e68222f8ca07a59bbfc1dbd8/packages/%40vue/cli-plugin-e2e-nightwatch/package.json#L31-L32

(I will update them in a batch later, so don't mind the chrome version being too old)

christian-bromann commented:
@sodatea thanks for the tip. Looks all good now.

sodatea commented:
Thanks!
